File: apps/api/src/middleware/subscriptionGuard.ts
import { subscriptionsEnabled } from "@makyschool/shared/constants";
import type { NextFunction, Response } from "express";
import { pool } from "../db/pool.js";
import type { TenantRequest } from "./tenant.js";

const EXEMPT_PREFIXES = ["/api/schools/setup"];

export async function requireActiveSubscription(
  req: TenantRequest,
  res: Response,
  next: NextFunction,
) {
  if (!subscriptionsEnabled()) {
    return next();
  }

  if (EXEMPT_PREFIXES.some((prefix) => req.path.startsWith(prefix))) {
    return next();
  }

  const schoolId = req.schoolId;
  if (!schoolId) {
    return res.status(400).json({ error: "Missing tenant context" });
  }

  const result = await pool.query<{ status: string; subscription_status: string }>(
    "SELECT status, subscription_status FROM schools WHERE id = $1 LIMIT 1",
    [schoolId],
  );

  const school = result.rows[0];
  if (!school) {
    return res.status(404).json({ error: "School not found" });
  }

  if (school.status === "suspended") {
    return res.status(403).json({ error: "School account is suspended", code: "SCHOOL_SUSPENDED" });
  }

  if (school.status === "setup") {
    return next();
  }

  if (school.subscription_status !== "active") {
    return res.status(402).json({
      error: "Subscription payment required for this term",
      code: "SUBSCRIPTION_REQUIRED",
    });
  }

  return next();
}

File: apps/api/app/middleware/subscription_guard.py
import uuid

import asyncpg
from fastapi import Depends, HTTPException, status

from app.db.pool import get_db
from app.middleware.tenant import get_tenant_and_user, resolve_tenant
from app.services.subscription import audit_school_subscription, subscriptions_enabled


async def require_active_subscription(
    school_id: uuid.UUID = Depends(resolve_tenant),
    conn: asyncpg.Connection = Depends(get_db),
) -> uuid.UUID:
    if not subscriptions_enabled():
        return school_id

    await audit_school_subscription(conn, school_id)

    school = await conn.fetchrow(
        "SELECT status, subscription_status FROM schools WHERE id = $1 LIMIT 1",
        school_id,
    )
    if not school:
        raise HTTPException(status_code=404, detail={"error": "School not found"})
    if school["status"] == "suspended":
        raise HTTPException(
            status_code=403,
            detail={"error": "School account is suspended", "code": "SCHOOL_SUSPENDED"},
        )
    if school["status"] == "setup":
        return school_id
    if school["subscription_status"] != "active":
        raise HTTPException(
            status_code=402,
            detail={
                "error": "Subscription payment required for this term",
                "code": "SUBSCRIPTION_REQUIRED",
            },
        )
    return school_id


async def require_tenant_with_subscription(
    ctx: tuple[uuid.UUID, dict] = Depends(get_tenant_and_user),
    conn: asyncpg.Connection = Depends(get_db),
) -> tuple[uuid.UUID, dict]:
    school_id, user = ctx
    if subscriptions_enabled():
        await audit_school_subscription(conn, school_id)
        school = await conn.fetchrow(
            "SELECT status, subscription_status FROM schools WHERE id = $1 LIMIT 1",
            school_id,
        )
        if school and school["status"] == "suspended":
            raise HTTPException(
                status_code=403,
                detail={"error": "School account is suspended", "code": "SCHOOL_SUSPENDED"},
            )
        if school and school["status"] != "setup" and school["subscription_status"] != "active":
            raise HTTPException(
                status_code=402,
                detail={
                    "error": "Subscription payment required for this term",
                    "code": "SUBSCRIPTION_REQUIRED",
                },
            )
    return ctx
